Mapping table column spec (#1220)
ni/nimble: packages/nimble-components/src/table/specs/table-column-specs/table-column-mapping.md

# Mapping Table Column

## Overview

> The `nimble-table-column-mapping` is a component that supports rendering specific number, boolean, or string values as mapped icons, spinners, or text. `nimble-table-column-icon` is a specialized version of `nimble-table-column-mapping` which defaults to having a minimal, fixed width. The actual mappings are defined by child elements `nimble-mapping-icon`, `nimble-mapping-spinner`, and `nimble-mapping-text`.

Reviewer: > which defaults to having a minimal, fixed width

My understanding is that we are going to be stricter than just defaulting the icon column to a fixed width -- we wouldn't even support resizing that column, so the "default" is really just the only behavior associated with the column. I think it would be good to be explicit about that here since you mention its sizing behavior. Alternatively, I think if you just want to discuss the sizing behavior below, then it might be better to not talking about it at all here and just saying something like _`nimble-table-column-icon` is a specialized version of `nimble-table-column-mapping` that supports showing a single icon or spinner._

Author: I hadn't thought we would prevent resizing. I figured we would support text column headers for icon columns, in which case a client would want to set the size such that the header was fully visible. You mention in a later comment that icon column headers should probably also be icons. That seems overly restrictive to me. I'd worry about the ability to communicate a column's meaning via an icon. I can go that direction for now, but I'd like to get confirmation from Leslie.

Author: I'm leaving another comment thread later in the PR open to track the issue of whether icon columns should be resizable.

> The `nimble-table-column-mapping` is a component that supports rendering specific number, boolean, or string values as mapped icons, spinners, or text. `nimble-table-column-icon` is a specialized version of `nimble-table-column-mapping` which defaults to having a minimal, fixed width. The actual mappings are defined by child elements `nimble-mapping-icon`, `nimble-mapping-spinner`, and `nimble-mapping-text`.

Reviewer: > The `nimble-table-column-mapping` is a component that supports rendering specific number, boolean, or string values as mapped icons, spinners, or text.

Technically the `nimble-table-column-mapping` isn't what supports a value being mapped to a icon/spinner, right? It maps to text, which may have an optional icon, right? I think it would be good to tighten up the language here a little bit.

Oh, I actually see below that you've called out _icon + text_ as out of scope. In that case, then I definitely think this wording should be updated since the mapping column doesn't support icons or spinners at all.

Author: I guess I had been thinking the `nimble-table-column-icon` was a convenience element, but hadn't made the jump to limiting the general mapping column to just text. That would make sense though, and would avoid ambiguity about which column type to use.

### Background

[Icon column type issue](https://github.com/ni/nimble/issues/1013)

[Boolean column type issue](https://github.com/ni/nimble/issues/1103)

### Features

-   Supported input:
    -   string
    -   number
    -   boolean
-   Supported output:
    -   Text
    -   Icon
    -   Spinner
    -   (empty)

### Non-goals

-   Non-Nimble icon support
-   Arbitrary icon colors
-   Hyperlink icons
-   Mixed text and icons
-   Non-icon, non-spinner Nimble components

---

## Design

Below is an example of how these elements would be used within a `nimble-table`:

```HTML
<nimble-table>
    <nimble-table-column-icon field-name="status">
        Status
        <nimble-mapping-icon key="fail" icon="nimble-icon-xmark" severity="error" label="Failed"></nimble-mapping-icon>
        <nimble-mapping-icon key="error" icon="nimble-icon-xmark" severity="error" label="Errored"></nimble-mapping-icon>
        <nimble-mapping-icon key="pass" icon="nimble-icon-check" severity="success" label="Passed"></nimble-mapping-icon>

>         <nimble-mapping-spinner key="running" label="Running..."></nimble-mapping-spinner>

Reviewer: nit: This has nothing to do with your spec, but I don't think we'd have a label that ended with "...". Could you update this example to remove the "..."?

```suggestion
        <nimble-mapping-spinner key="running" label="Running"></nimble-mapping-spinner>
```

    </nimble-table-column-icon>
    <nimble-table-column-mapping field-name="errorCode" data-type="number">
        Error Code
        <nimble-mapping-text key="1" label="A bad thing happened"></nimble-mapping-text>
        <nimble-mapping-text key="2" label="A worse thing happened"></nimble-mapping-text>
        <nimble-mapping-text key="3" label="A terrible thing happened"></nimble-mapping-text>
    </nimble-table-column-mapping>
    <nimble-table-column-icon field-name="archived" data-type="boolean">
        Archived
        <nimble-mapping-icon key="true" icon="nimble-icon-database" label="Archived"></nimble-mapping-icon>
    </nimble-table-column-icon>
</nimble-table>
```

Note that the `key` attribute values are always given as strings. In the case of boolean or number mappings, this value is parsed to the appropriate type, so that it can properly be compared to the values from the row records.

> When none of the given mappings match the record value for a cell, that cell will be empty. Alternatively, we could provide support for marking a mapping with a `default` attribute that would cause it to be chosen when no other values matched. This would be equivalent to the `placeholder` configuration we provide on `nimble-table-column-text` and `nimble-table-column-anchor`. I don't think we need to support this initially.

Reviewer: Is there a reason to not support a placeholder initially for the `nimble-table-column-mapping` column? I could see a client wanting to specify some sort of "default" value if the value doesn't match one of the enumerated `key` values.

Author: No real reason. Frankly, I just anticipate it being pretty tricky to implement in a template with just `repeat` and `when`, but maybe I'm wrong. I'll include it in the plan.

Author: Turned out not to be that tricky.

If multiple mappings in a column have the same key, an error will be thrown.

> If an invalid `icon` value is passed to `nimble-mapping-icon`, an error will be thrown.

Reviewer: I think we should be more specific about the validation being done. Specifically, some questions I have are:
- What makes an icon valid vs invalid? Is a valid icon any element that has `Icon` as its base class?
- Why throw an error? We've generally made the statement that setting an attribute on an element shouldn't throw an error because there isn't a good way to catch that error. Also, I could see the validation code being async, [similar to the column filtering code in the table](https://github.com/ni/nimble/blob/486a75bd2be240c186a627a6ceb93f433b14a478/packages/nimble-components/src/table/index.ts#L565), because we don't know if all custom elements have been defined yet. Should this valid/invalid state really bubble up to the table's `validity` object in some way?

Author: > What makes an icon valid vs invalid? Is a valid icon any element that has Icon as its base class?

That seems like an accurate way to do it. I'll update the spec.

> Why throw an error? ... Should this valid/invalid state really bubble up to the table's validity object in some way?

I welcome some guidance here. Naively, it sounds like I should set some flags in the table's validity object instead of throwing errors. Are there any reasons I shouldn't?

Reviewer 2: Seems like there is a lot of validation needed for this column type compared to others so far it may be worthwhile to figure out something now. I'll say anecdotally having that added to the table made it a lot easier to reason about the behavior of error conditions.

API-wise, just brainstorming, I think something like:
```ts
abstract class Validator {
    abstract isValid()
}

class TableValidator extends Validator {
   private invalidColumnConfiguration // true if one or more invalid columns
}

class IconColumnValidator extends Validator {
}

class TableColumn {
    abstract readonly columnValidity: Validator;
}
```
Let's the table not need to know about specific column details and columns can determine there own properties. Just one idea off-the-top.

Reviewer: I agree that correct configuration is pretty fundamental to getting these new column types to behave properly, so we shouldn't defer figuring validation out. However, I'm almost always in favor of making smaller PRs -- just because we say we shouldn't defer the work doesn't mean it has to all be part of one giant PR.

Another note, is that after thinking through this a bit more, I've realized that there are really two different types of validation for a column - validation that the data records match the types the columns are expecting and validation of the configuration of the column & any children it has. The issue I linked in my comment was thinking about the first type of validation. The majority of the validation you need to do here is the second type of validation, which doesn't need access to the data records and can be determined entirely by the column with no additional information (i.e. doesn't need to be a function that gets called on the column). I haven't fully thought through @rajsite's example API, but I think we can potentially leverage `@observable` properties on the column to help communicate a column's validity back up to the table in a way that isn't possible when the table's data is required to validate a column.

Author: I'm still unsure how much detail we want to surface with regards to invalid column configuration. We could do what Milan had in his suggestion and have just a single flag for any column being invalid, but it would be nice if we could give more specifics. Should we let the table validator store specific validation error strings that it collects from individual columns? Should columns have separate validity objects that the client has to read from? (That seems overly inconvenient.)

> we can potentially leverage @observable properties on the column...

Is there something significant about having `@observable` properties over just validity-checking functions?

Reviewer: >> we can potentially leverage `@observable` properties on the column...

> Is there something significant about having `@observable` properties over just validity-checking functions

Yes, there is a fairly significant architectural difference between `@observable` properties vs validity-checking functions -- who is responsible for knowing that the validity may have changed & needs to be rechecked. With `@observable` properties, the column can own all of its own logic for what could make it invalid (e.g. specific attributes changing, child items changing, etc) and tell the table when the state has changed. With validity-checking functions, the table somehow has to know that the validity of a given column might have changed and so the function needs to be called again. Therefore, I think with `@observable` properties it would be pretty straightforward to have the column own its own validity and not leak implementation details into the table. It's possible this can be done with a function, but I think it wouldn't be as clear and/or clean.

Author: Sure, I was just thinking that with the current scheme of the validation object only being updated when the `validate()` function is called, column validation functions could just be called when all the existing validation functions get called. But `@observable`s facilitate updating validation at other times, which may be useful.

Author: I've updated the table column HLD with the plan for column validation.

> `nimble-table-column-icon` supports only `nimble-mapping-icon` and `nimble-mapping-spinner` as mapping elements. Any others will result in an error being thrown. `nimble-table-column-mapping` supports all mapping elements.

Reviewer: Why does the `nimble-table-column-mapping` need to support `nimble-mapping-icon` and `nimble-mapping-spinner` as children? I would have thought we only want `nimble-mapping-text` its children.

Author: Now restricting to just text.

**Alternatives:**

An earlier version of this spec proposed mapping elements with `template` elements as content instead of relying on attribute configuration. The template element would define the mapped html to render. We would impose restrictions on the types of supported elements that could be provided in the template.

Pros:

-   It would not require updates to the API if we needed to support new types of mapped content (e.g. icon with text), or if the mapped content itself got new configuration options (e.g. a scaling factor for icons).
-   Undefined element types caught at compile time.

Cons:

-   Verbose. Requires user to create `template` element and wrap text in `span`s for styling purposes.
-   Requires difficult validation to ensure only supported elements are present in the `template`.
-   Could allow users to provide inline styling.
-   Blazor: cannot put Blazor components inside `template`--must use raw Nimble elements without type safety

### API

#### Icon column element:

_Component Name_

-   `nimble-table-column-icon`

_Props/Attrs_

-   `field-name`: string

> -   `data-type`: 'string' | 'boolean' | 'number' (defaults to 'string')

Reviewer: Why does the client need to specify the `data-type`? Shouldn't the data know its own type?

Author: I guess that is true.

> -   `pixel-width`: number (defaults to minimum width supported by table)

Reviewer: I don't think we should let someone resize the icon column.

Author: Ok, done.

_Content_

> -   column title (text)

Reviewer: The column title is slotted content, so there isn't a requirement that the column title can only contain text. More over, I think the guidance will be that an icon is used as the header content for icon columns so that the column doesn't have to be resizable but the header content can be seen.

Author: From what Leslie said in the meeting today, we do want to support text column headers for icon columns. It sounds like there's some concern about letting icon columns be resizable. I'm not sure I understand the aversion. For the time being, I'll work under the assumption that we will support resizing icon columns.

-   1 or more `nimble-mapping-icon` or `nimble-mapping-spinner` elements

#### General mapping column element:

_Component Name_

-   `nimble-table-column-mapping`

_Props/Attrs_

-   `field-name`: string

> -   `data-type`: 'string' | 'boolean' | 'number' (defaults to 'string')

Reviewer: Similar to above, what's the rationale for needing to specify the `data-type` on the API for the column?

Author: I guess we don't need to.

> -   `pixel-width`: number (set to the desired fixed column width, else will have a fractional width)

Reviewer: My guess is that we don't want to support `pixel-width` on the mapping column. The `nimble-column-text` column only supports `fractional-width`, so I think that's what we should do here as well unless there is a compelling reason to do something different.

Author: Yes, now that I'm not trying to support icons in the mapping column, this is a logical consequence.

-   `fractional-width`: number (defaults to 1)
-   `min-pixel-width`: number (defaults to minimum supported by table)

_Content_

> -   column title (text)

Reviewer: As mentioned above, the column title doesn't have to be strictly text since it is slotted content.

Author: Changed.

-   1 or more `nimble-mapping-*` elements

#### Mapping element (icon):

_Component Name_

-   `nimble-mapping-icon`

_Props/Attrs_

-   `key`: string (will also have a private, typed version of this property)
-   `icon`: string - name of the Nimble icon element
-   `severity`: string - one of the supported enum values. Controls color of the icon.

> -   `label`: string - localized value used as the accessible name and `title` of the icon

Reviewer: It might be good to mention here that the label will also be used for the grouped row if the table is grouped by this column.

Author: Done

#### Mapping element (spinner):

_Component Name_

-   `nimble-mapping-spinner`

_Props/Attrs_

-   `key`: string (will also have a private, typed version of this property)

> -   `label`: string - localized value used as the accessible name and `title` of the spinner

Reviewer: Similar to the icon, we should mention here that the label will also be used on the grouped row.

        <slot></slot>
    </span>
</template>
```

Cell view:

The cell view relies on the matched mapping to provide a template to render.

```HTML
${repeat(x => (x.column as TableColumnMapping).mappings,
    html<TableColumnMapping, TableColumnMappingCellView>`
        ${when((x, c) => x.typedKey === (c.parent as TableColumnMappingCellView).cellRecord.value,
            html<TableColumnMapping>`${x => x.cellViewTemplate}`
        )}
    `
)}
```

Group header view:

Similarly, the group header view relies on the matched mapping to provide a template to render.

Note the following requires that `TableColumnMappingGroupHeaderView` has a reference to the column with which it is associated. This is needed to enumerate the column's mapping elements.

```HTML
${repeat(x => (x.column as TableColumnMapping).mappings,
    html<TableColumnMapping, TableColumnMappingHeaderView>`
        ${when((x, c) => x.key === (c.parent as TableColumnMappingHeaderView).groupHeaderValue,
            html<TableColumnMapping>`${x => x.groupHeaderViewTemplate}`
        )}
    `
)}
```

#### `nimble-mapping-*`

```HTML
<template slot="mapping"></template>
```

#### `nimble-mapping-icon`

Cell view template:

> ```

Reviewer: I'm guessing filling this in is a TODO?

Author: Whoops. Meant to come back and fill that in.

```

### Grouping

> Will support grouping by the record value. The grouping header will display the rendered icon/spinner/text. In the case of an icon/spinner, it will also be followed by the `label` text. This will disambiguate cases where multiple record values map to the same icon (assuming the labels are different). Text in a grouping header will be ellipsized and gain a tooltip if there is not enough room to display it all.

Reviewer: Can you make this first sentence a complete sentence?

```suggestion
Grouping will be based on the record value. The grouping header will display the rendered icon/spinner/text. In the case of an icon/spinner, it will also be followed by the `label` text. This will disambiguate cases where multiple record values map to the same icon (assuming the labels are different). Text in a grouping header will be ellipsized and gain a tooltip if there is not enough room to display it all.
```

Author: Changed

### Sorting

Sorting will be based on the record value. For icons, if multiple values map the to same icon, it is possible that sorting will result in the instances of a certain icon not being all together in one span of rows. The user may think they should be all together, but this is a corner case that we can't/shouldn't do anything about.

### Sizing

`nimble-table-column-icon` will support only a fixed width. We will introduce a new mixin for fixed-width support that exposes a `pixel-width` property. The default value will be the minimum supported by the table, which is still significantly larger than the width of an icon.

`nimble-table-column-mapping` will support fixed or fractional widths. If `pixel-width` is set, the column will have a fixed width, otherwise it defaults to a fractional width of 1. The client may configure `fractional-width` and/or `min-pixel-width`.

### Angular integration

Angular directives will be created for the column components and the mapping components. No component has form association, so a `ControlValueAccessor` will not be created.

### Blazor integration

Blazor wrappers will be created for the components.

### Visual Appearance

> The cell view (and group header view) will be responsible for styling the templates returned by the mappings. This will include alignment and spacing (`--ni-nimble-small-padding`). Unfortunately, because the mappings cannot provide CSS to go with the templates they return, some implementation knowledge will leak from the mappings to the cell/group header views. For example, we must set `flex-shrink: 0` on all elements so that icons do not get smaller in a group header when the adjacent label takes up all the space.

Reviewer: I'm not sure I quite follow all the reasoning here? Isn't the cell/group header view specific for a given column type? For example, having the cell/group views for the icon column know that it contains an icon that shouldn't be shrunk doesn't seem like leaking implementation. Am I missing something?

Author: Well, `nimble-table-column-icon` has a cell view template, and it has styling for that template. However the styling makes assumptions about things that are coming from the `nimble-mapping-icon` element. I.e. this is the column cell view's template, and it contains no icons:

```HTML
${repeat(x => (x.column as TableColumnMapping).mappings,
    html<Mapping, TableColumnMappingCellView>`
        ${when((x, c) => x.typedKey === (c.parent as TableColumnMappingCellView).cellRecord.value,
            html<Mapping>`${x => x.cellViewTemplate}`
        )}
    `
)}
```
It's the `x.cellViewTemplate` in the last line that provides the actual content, including the icons. It feels kind of wrong for the column cell view's CSS to provide styling for that.

# Mapping Table Column

## Overview

The `nimble-table-column-mapping` is a component that supports rendering a specific number, boolean, or string value as mapped, predefined, html fragments (with content restrictions).

### Background

[Icon column type issue](https://github.com/ni/nimble/issues/1013)

[Boolean column type issue](https://github.com/ni/nimble/issues/1103)

### Features

-   Supported input:
    -   string
    -   number
    -   boolean
-   Supported output:
    -   Text
    -   One or more Nimble icons (with colors via icon severity property)
    -   Nimble spinner

>     -   Mixed icons and text

Reviewer: Will nimble automatically space multiple items, such as icon + text? Or, do we need to make sure to have documentation on the recommended spacing between those items?

Author: I think our component will have to do the spacing for them. I've added a remark to this effect. 

The client can't set any styling that would affect the elements we render in the cell view's shadow DOM. (My proposed implementation renders them using the `template`'s `innerHTML`. We're not slotting elements that exist in the light DOM.)

    Mappings of column types:

    - icon column:
        - number -> icon/spinner
        - string -> icon/spinner
        - boolean -> icon/spinner
    - enum column:
        - number -> text
        - string -> text
        - number -> icon/spinner
        - string -> icon/spinner
    - boolean column:
        - boolean -> text
        - boolean -> icon/spinner

    An icon column would cover half the use cases of enum and boolean columns. The remaining use cases would be covered by something like a mapped text column. This could be confusing to distinguish from our text column. Additionally, if we are going to have a column that maps numbers/strings/booleans to text, it might as well generalize to icons as well. In which case, we don't need a distinct icon column.

### API

#### Column element:

_Component Name_

-   `nimble-table-column-mapping`

_Props/Attrs_

-   `field-name`: string

_Content_

-   column title (text)

> -   1 or more `nimble-mapping-*` elements

Reviewer: Do you think these elements would ever be used in a component other than the `nimble-table` or within a column other than the `nimble-table-column-mapping`? If we don't think they'd be used outside of that context, perhaps the name should reflect that. My first thought is `nimble-table-column-mapping-value`, but I could also see the prefix "nimble-table-column" being misleading since it isn't a table column.

                    html<HTMLTemplateElement>`${x => new ViewTemplate(x.innerHTML, [])}`
                )}
            `
        )}
    `
)}
```

For each mapping, if the key matches the value for this row, get the mapping's contained `template` and render its `innerHTML`.

#### `nimble-mapping-string/number/boolean`

```HTML
<template slot="mapping">
    <slot
        ${slotted({ property: 'mappedTemplate', filter: x => x instanceof HTMLTemplateElement })}
    >
    </slot>
</template>
```

### Grouping

Will support grouping by the record value. The grouping header will display the rendered html mapped to that value. If multiple different values map to the same rendered html (e.g. icon), there may be separate groups that visually look like they should be combined into one group. We could instead render the record value into the header, but that value is not otherwise exposed to the user. It may not even have much meaning (e.g. some numeric code). I think it is more reasonable to display the rendered html.

### Sorting

Sorting has the same issue as grouping, where visually identical cells could end up not sorted contiguously. Again, I believe this is reasonable.

### Sizing

> We will support fixed or fractional widths for this column type.

Reviewer: To better support mapping columns with only icon mappings, should we export a constant (or a token?) that makes it easy for a client to create an icon column with the recommended fixed size?

Author: I'm just now looking at how we support fixed-width columns. Do I need to define a `pixel-width` attribute on the column type and sync that with `columnInternals.currentPixelWidth`? The example in the spec shows setting `columnInternals.currentPixelWidth` in the component's constructor, but that doesn't seem applicable to this situation. But to answer your question, I suppose it would be helpful to export a constant for a single-icon column width.

Author: So it sounds like we need to define a new mixin to expose the public API for setting fixed widths on columns. It would do basically what I suggested in my last comment. Also, for icon-only columns, it sounds like the default minimum column width is significantly more than the width of an icon. So maybe we don't need to have a token for the column width. We can just document to set the width to 1 and let the minimum govern the width.

Reviewer 3: I'm not entirely sure what you mean by "set the width to 1 and let the minimum govern the width". Just so we're on the same page about things, the default behavior for _any_ column, regardless of the mixins they have applied, will be to behave as a fractional width column of `1fr`. If we want an icon column to default to a pixel width behavior, then it is _required_ to set the `currentPixelWidth` to a non-undefined value.

This is arguably a problem with how the current API is designed. That is, if we had something like a `FixedWidthColumn` mixin that itself did _not_ extend the `FractionalWidthColumn`, then unless a non-null default was provided for the `pixelWidth` (which perhaps it should be given the same default as `minPixelWidth`), then it would still behave like a fractional width column by default.

Author: I had not assumed we should make this column type use a fixed width by default, but I can see the argument for doing so. It could result in truncated text (by default) for columns that map values to text, but it's arguably better to expect clients to override the default in those cases than have to set a fixed width in all the other cases (i.e. with just icons/spinners).

Author: Now we're going to have a separate column type for icons which only supports fixed width. The other, general mapping type will support both, but default to fractional.

### Angular integration

> An Angular directive will be created for the component. The component will not have form association, so a `ControlValueAccessor` will not be created.

Reviewer: I'm guessing directives will be created for the mapping elements as well? It would be good to be explicit about that here.

Author: Made that explicit.

### Blazor integration

A Blazor wrapper will be created for the component. _Initial attempts at prototyping resulted in exceptions being thrown and no icons being rendered. This is an open issue._

### Visual Appearance

Appearance can vary based on mapped html, but can only include text, icons, and Nimble spinner.

---

## Implementation

_Important aspects of the planned implementation with careful consideration of web standards and integration._

_Highlight any alternative implementations you considered in each section._

_If you think a section doesn't apply or don't know what to write, please DO NOT delete it. Either mark it "N/A" or leave it blank and the Nimble team can help you fill it in._

### States

N/A

### Accessibility

Text, icons, and spinner are not interactive and cannot receive keyboard focus.

The cell will not introduce any items into the accessibility tree other than the elements in the mapped template. Those items will already have appropriate accessibility tree representations.

> We will instruct clients to provide an accessible name for each icon/spinner, either via the `title` or `aria-label` attribute. `aria-labelledby` is not an option, because it requires defining an `id` which will not be unique when duplicated accross multiple rows.

Reviewer: Setting `title` does not create a label in the accessibility tree. Are there alternatives that would make this easier for clients? For example, could the icon mapping element have an attribute that was the label, and nimble would set that as the `aria-label` and as the `title`?

Author: > Setting title does not create a label in the accessibility tree.

I believe it does (though, as a "fallback"). As [this W3C doc](https://www.w3.org/WAI/ARIA/apg/practices/names-and-descriptions/) states, "The title attribute may be used as the element's fallback accessible name."

> could the icon mapping element have an attribute that was the label...

I initially had this, but the question becomes which element(s) to set that on. We could set it on a `div` wrapping the whole thing, but that doesn't actually provide an accessible name to the wrapped elements. If we're directing the client to set `description` on our mapping element, I think we might as well just direct them to set the `title` or `aria-label` on the appropriate element instead.

Reviewer 2: > I believe it does (though, as a "fallback"). As this W3C doc states, "The title attribute may be used as the element's fallback accessible name."

I don't have an explicit alternative proposal yet. I'll just point out that the same cited doc says not to rely on title as a fallback:

> Rule 4: Avoid Browser Fallback
> When authors do not specify an accessible name using an element or attribute that is intended for naming, browsers attempt to help assistive technology users by resorting to fallback methods for generating a name. For example, the HTML title and placeholder attributes are used as last resort sources of content for accessible names. Because the purpose of these attributes is not naming, their content typically yields low quality accessible names that are not effective.

We had a similar discussion in the Web Chapter meeting on the [Hijacking Screenreaders with CSS presentation](https://www.youtube.com/watch?v=kk-03n3Y4WI) (good watch, we have hit many of the issues it talks about) where in regards to `text-transform: uppercase`:

> Maybe we should strive for is minimizing the heuristics we rely on. I bet the browsers will each consistently report it as the text ADD in the accessibility tree (we can verify that in chrome and firefox devtools, I assume Safari has an accessibility tree inspector too?), but we know that's not what we want to be in the tree. So even if heuristics work pretty well today for certain acronyms in certain languages, we probably don't want to rely screen readers correcting the wrong thing being shown in the accessibility tree.

Author: Good point. In that case, I think we would document that the user should set `aria-label` and optionally `title` as well on the element(s) in their template.

Author: Now we will be setting `aria-label` ourselves, based on the provided `label` value.

```HTML
<nimble-table>
    <nimble-table-column-icon field-name="status">
        Status
        <nimble-mapping-icon key="fail" icon="nimble-icon-xmark" severity="error" label="Failed"></nimble-mapping-icon>
        <nimble-mapping-icon key="error" icon="nimble-icon-xmark" severity="error" label="Errored"></nimble-mapping-icon>
        <nimble-mapping-icon key="pass" icon="nimble-icon-check" severity="success" label="Passed"></nimble-mapping-icon>
        <nimble-mapping-spinner key="running" label="Running"></nimble-mapping-spinner>
    </nimble-table-column-icon>
    <nimble-table-column-mapping field-name="errorCode">
        Error Code
        <nimble-mapping-text key="1" label="A bad thing happened"></nimble-mapping-text>
        <nimble-mapping-text key="2" label="A worse thing happened"></nimble-mapping-text>
        <nimble-mapping-text key="3" label="A terrible thing happened"></nimble-mapping-text>
    </nimble-table-column-mapping>
    <nimble-table-column-icon field-name="archived">
        Archived
        <nimble-mapping-icon key="true" icon="nimble-icon-database" label="Archived"></nimble-mapping-icon>
    </nimble-table-column-icon>
</nimble-table>
```

> Note that the `key` attribute values are always given as strings. In the case of boolean or number mappings, this value is parsed to the appropriate type, so that it can properly be compared to the values from the row records.

Reviewer 4: Can you elaborate on the recommendation that `key` attribute values are always strings?
1. what does this look like in client code? If I have an integer enum that I want to map to icons, do I have to explicitly call `toString()` in the template?
2. what alternatives did we consider and why did we choose this approach? Some that come to mind:
    1. mapping elements expose multiple attributes `key-string`, `key-numeric`, `key-boolean`. Nimble could enforce that only one attribute is used within a mapping column.
    2. `key` attribute accepts values of type `string | number | boolean`. Nimble could enforce that all values are the same type within a mapping column.
    3. different mapping elements for different input types. `nimble-mapping-text-string`, `nimble-mapping-text-numeric`, `nimble-mapping-text-boolean`, `nimble-mapping-icon-string`, ...

Author: I think maybe the point I was trying to make is too obvious, or I am just confused. I simply meant that in html, attribute values are always specified in quotes (i.e. as a string): `<element attr="value" ...`. The `key` property will be typed as `string | number | boolean`. I guess I'm not clear what type you get for `key="true"`--does typescript treat that value as the string "true" or the boolean `true`? I was assuming we would have to run type conversions from string to bool/number.

Reviewer 4: Gotcha, I think I was also confused so thanks for the clarification.

Reviewer: I think I'm confused now. The attribute will always be specified as a string but the property will be `string | number | boolean`? When will the conversion happen? Will it happen when you check the value or will it happen as part of a custom attribute to property converter that we write?

Reviewer 2: With that direction I'd expect the column to do the conversion, the column is the one that know the target type from the field, not the mapping itself.

I think in this case we should also configure the mode of the key attribute to be `mode: 'fromView'`: https://www.fast.design/docs/fast-element/defining-elements#customizing-attributes

Since it would be ambiguous when assigned back to the attribute what the actual type of the serialized string is.

Reviewer 2: Actually.. is that kind of pattern possible to represent in a Blazor component?

Would three nullable keys be less fiddly? `numeric-key`, `string-key`, `boolean-key`? It's spiritually similar to the [`input` element valueAsNumber, valueAsDate, etc](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement).

Author: I confirmed that typing `key` as `string | number | boolean` doesn't really do anything for me. When I do `typeof mapping.key`, I always get `string`. So I think I might as well just type it as such.

My current implementation doesn't convert the value until the last moment, when it is trying to compare against the `cellView.cellRecord.value`. It's at that point that I can get the `typeof` the record value and convert the mapping's `key` value accordingly:
```ts
public isMappingFor(value: TableFieldValue): boolean {
  const dataType = typeof value;
  if (dataType === 'number') {
      return value === nullableNumberConverter.fromView(this.key);
  }
  if (dataType === 'boolean') {
      return value === booleanConverter.fromView(this.key);
  }
  if (dataType === 'string') {
      return value === this.key;
  }
  return false; // value is null or undefined
}
```

Reviewer 2: > I confirmed that typing key as string | number | boolean doesn't really do anything for me. When I do typeof mapping.key, I always get string

TypeScript types don't exist at runtime. How the property behaves depends on if you do property assignment or attribute assignment. FAST doesn't reject property assignments that are invalid, so assigning `el.key = 5` will assign a `typeof this.key === 'number'` value. The typing is for a user using TypeScript to limit what can be assigned. For attributes it depends on the converter used what ends up in the proeprty from an attribute assignment.

If we make it only string then they have to do something like `el.key = String(5)` for TypeScript to not complain.

Author: > Actually.. is that kind of pattern possible to represent in a Blazor component?
> 
> Would three nullable keys be less fiddly? numeric-key, string-key, boolean-key?

I think you're right. For the sake of Blazor, I think we have to have three separately typed attributes. I considered whether we should also have a multi-typed `key` attribute to simplify the API for vanilla html and Angular, but I think it would probably just confiuse things more.

Reviewer 3: > Actually.. is that kind of pattern possible to represent in a Blazor component?
> 
> Would three nullable keys be less fiddly? `numeric-key`, `string-key`, `boolean-key`? It's spiritually similar to the [`input` element valueAsNumber, valueAsDate, etc](https://developer.mozilla.org/en-US/docs/Web/API/HTMLInputElement).

In Blazor, we would want this to be generic (ideally). So on the Blazor `NimbleMappingText` component, if I wanted my key to be an integer (because that's the type of the data), I would have:

```html
<NimbleTableColumnMapping>
    <NimbleMappingText TValue=int Key="1"></NimbleMappingText>
    <NimbleMappingText TValue=int Key="2"></NimbleMappingText>
</NimbleTableColumnMapping>
```

It would be great if `NimbleTableColumnMapping` could be the generically typed object that would then give the `NimbleMappingText` objects within an implicit typing, but we can't do that. This means that the user could also give `NimbleMappingText` objects different typing, which would have to be flagged as invalid by the table. I think we could implement the component side in a way like Milan suggests, where the component will set either `numeric-key`, `string-key`, or `boolean-key` based on the type of `TValue`. The component would throw if an invalid type was given to `TValue`.

Author: Does `NimbleMapping*` have to be generic if we have distinct properties `key-number`, `key-string`, and `key-boolean`?

Reviewer 3: For Blazor we wouldn't want those distinct properties on the Blazor API surface (that is what generics are for). We can hide the `key-number`, `key-string`, and `key-boolean` attributes as implementation details (set in the component .razor code conditionally).

I'll admit that this is an example of consciously diverging the web component and Blazor APIs, but I feel that this is a good example of us changing to adopt an expected API for the platform (similar to having a property for the data instead of a method).

Author: That brings me back to the question of whether we actually need three distinctly typed key properties. If the Blazor component is going to have a single, generically typed key property, can we have a single key property (of type `number | string | boolean | null`) on the underlying component?

Reviewer 3: But isn't the issue with converting the key value to the appropriate type? If you have the three distinct properties there is no confusion. How does a particular mapping column know that its `key` should be a `number` versus a `string`?

Author: We've come around to saying we're going to put a `data-type` attribute on the column, which means we don't really need the `key` property to specify its own type. Going back to a single `key` attr/prop.

When none of the given mappings match the record value for a cell, that cell will be empty. Alternatively, if one of the mappings has the `default` attribute, it will match when no other mappings have. This is equivalent to the `placeholder` configuration we provide on `nimble-table-column-text` and `nimble-table-column-anchor`.

If multiple mappings in a column have the same key, an error flag will be set on the table's validity object.

> If an invalid `icon` value is passed to `nimble-mapping-icon`, an error flag will be set on the table's validity object. An invalid `icon` value is any element that does not derive from `Icon`.

Reviewer 4: nit: We should probably also call it invalid if the element can't be resolved at all (e.g. if the client refers to an icon that they forgot to import).

```HTML
${repeat(x => (x.column as TableColumnMapping).mappings,
    html<Mapping, TableColumnMappingHeaderView>`
        ${when((x, c) => x.key === (c.parent as TableColumnMappingHeaderView).groupHeaderValue,
            html<Mapping>`${x => x.groupHeaderViewTemplate}`
        )}
    `
)}
```

#### `nimble-mapping-*`

```HTML
<template slot="mapping"></template>
```

#### `nimble-mapping-icon`

Cell view template:

```HTML
<${this.icon}
    title="${x => x.label}"
    aria-label="${x => x.label}"
    severity="${x => x.severity}">
</${this.icon}>
```

Group header view template:

```HTML
<${this.icon}
    title="${x => x.label}"
    aria-label="${x => x.label}"
    severity="${x => x.severity}">
</${this.icon}>
<span
    ${ref('span')}
    @mouseover="${x => {
        x.isValidContentAndHasOverflow = !!x.label && x.span!.offsetWidth < x.span!.scrollWidth;
    }}"
    @mouseout="${x => {
        x.isValidContentAndHasOverflow = false;
    }}"
    title=${x => (x.isValidContentAndHasOverflow ? x.label : null)}>
    ${x => x.label}
</span>`;
```

### Grouping

> Grouping will be based on the record value. The grouping header will display the rendered icon/spinner/text. In the case of an icon/spinner, it will also be followed by the `label` text. This will disambiguate cases where multiple record values map to the same icon (assuming the labels are different). Text in a grouping header will be ellipsized and gain a tooltip if there is not enough room to display it all.

Reviewer 4: How will group headers work for values that don't have a mapping? If a mapping is marked as `default` will its group header be used? Will all those values be in one group or in separate groups with the same header? What if no mapping is marked as `default`?

Author: I feel pretty strongly that we should show the raw (unmapped) data value in those cases. It seems unhelpful and confusing to show either a blank or the default mapping, specifically when there are multiple, separate, unmapped values showing the same header. I've update the spec to this effect. I realize this might be controversial, because we generally should not be mixing the display of source values and mapped values. I just think it's the more practical approach in this case.

Reviewer 4: I am ok with this if it doesn't require us to do any contortions in the implementation to make that information available where it is needed. But thinking it through a little more, this is a bit of an error case: it indicates that the client didn't configure the column correctly for the data that it is being asked to display. Since it's related to data rather than configuration I don't think it merits triggering a validation error. But it will hopefully not be a case that users see if our clients do their jobs right.

### Sorting

> Sorting will be based on the record value. For icons, if multiple values map the to same icon, it is possible that sorting will result in the instances of a certain icon not being all together in one span of rows. The user may think they should be all together, but this is a corner case that we can't/shouldn't do anything about.

Reviewer 4: Could you think through which sorting mode (e.g. locale-aware or basic) we'd use for each data type? 

I wonder if users would find it confusing that we sort by record value if the mapped string values would produce a different sort order. Is there any way we could provide a custom sorting mode for mapped string values?

Author: Please read through the update I've made to that section of the spec. TLDR: I think we should just use a basic sort regardless of the data type.

> Is there any way we could provide a custom sorting mode for mapped string values?

No, TanStack only supports sorting by a value in the record, which the mapped value is not. Also, if we want to support server-side sort, that also conflicts with sorting by a mapped value.

### Blazor integration

Blazor wrappers will be created for the components.

### Visual Appearance

The cell view (and group header view) will be responsible for styling the templates returned by the mappings. This will include alignment and spacing (`--ni-nimble-small-padding`). Unfortunately, because the mappings cannot provide CSS to go with the templates they return, some implementation knowledge will leak from the mappings to the cell/group header views. For example, we must set `flex-shrink: 0` on all elements so that icons do not get smaller in a group header when the adjacent label takes up all the space.

---

## Implementation

### States

N/A

### Accessibility

Text, icons, and spinner are not interactive and cannot receive keyboard focus. These items already have proper accessibility roles, and we will set accessible names (`aria-label`) based on the `label` value provided by the client.

### Globalization

All text will be provided by the client and is expected to be localized.

### Security

N/A

### Performance

> N/A

Reviewer 4: It wouldn't hurt to do some manual performance testing to ensure that scrolling performance is still good with the icon column.

-   Blazor: cannot put Blazor components inside `template`--must use raw Nimble elements without type safety

### API

#### Icon column element:

_Component Name_

-   `nimble-table-column-icon`

_Props/Attrs_

-   `field-name`: string
-   `pixel-width`: number (set to the desired fixed column width, else will use a default fixed width)

_Content_

-   column title (text or icon)
-   1 or more `nimble-mapping-icon` or `nimble-mapping-spinner` elements

#### General mapping column element:

_Component Name_

-   `nimble-table-column-mapping`

_Props/Attrs_

-   `field-name`: string
-   `fractional-width`: number (defaults to 1)
-   `min-pixel-width`: number (defaults to minimum supported by table)

_Content_

-   column title (text or icon)
-   1 or more `nimble-mapping-text` elements

#### Mapping element (icon):

_Component Name_

-   `nimble-mapping-icon`

_Props/Attrs_

-   `key`: string | number | boolean
-   `icon`: string - name of the Nimble icon element
-   `severity`: string - one of the supported enum values. Controls color of the icon.
-   `label`: string - localized value used as the accessible name and `title` of the icon. Will also be displayed in the group header.

> -   `default`: boolean - presence causes this mapping to be used when no others match the value

Reviewer: Is there going to be validation that only one mapping element is specified as the default?

Author: Yes, thanks for reminding me to list that.

    @mouseover="${x => {
        x.isValidContentAndHasOverflow = !!x.label && x.span!.offsetWidth < x.span!.scrollWidth;
    }}"
    @mouseout="${x => {
        x.isValidContentAndHasOverflow = false;
    }}"
    title=${x => (x.isValidContentAndHasOverflow ? x.label : null)}>
    ${x => x.label}
</span>`;
```

### Grouping

Grouping will be based on the record value. The grouping header will display the rendered icon/spinner/text. In the case of an icon/spinner, it will also be followed by the `label` text. This will disambiguate cases where multiple record values map to the same icon (assuming the labels are different).

For values that do not match any mapping, we will display the raw data value. While this introduces inconsistency, it seems preferable to the alternative, which is having multiple, separate groupings with a blank header (well, with just the item count in parens). Even in the case where there is a default mapping, we would still end up with separate groups with the identical default mapped icon and/or text, which is just a bad.

Text in a grouping header will be ellipsized and gain a tooltip if there is not enough room to display it all.

### Sorting

> Sorting will be based on the record value. For boolean and number values, a basic sort (just using basic comparison/equality operators) is the clear choice. For string values, it is less clear. In the case where the strings are enum values, they are likely to be non-localized, English strings. In that case, even if there is a semantically meaningful sort order (i.e. "NOT_STARTED" < "RUNNING" < "DONE"), we can't sort that way. Settling for a basic sort in this case seems just as reasonable as anything else. If the strings are localized, it's more likely the client would use a text column instead. If we did want to support a different sorting method for strings than booleans/numbers, it would be awkward to implement. We would have update the column's sort method when the data changes, and the column would have to determine the data type for its field, which might require looking at multiple records (e.g. if the first didn't define that field). For these reasons, we will just use a basic sort for all supported data types (string, boolean and number).

Reviewer 2: It sounds like the concern here is having a clear way for users to express the sorting of the mappings.
That is already expressed one way. We could just use the order of the mapping elements in the DOM.

Author: That's an interesting idea. The current implementation only supports specifying the sort function (by an enum value) to one of a couple presets. I don't know if there's a good way to make it support using a function bound to an instance of my column (which I would need to access the ordered mappings). I'll have to think a bit.

Reviewer 2: We were reset but looks like this thread is still open. Do we know if there are current expected use-cases of needing to sort mapped values? Maybe we can discuss this at stand-up?

Author: After discussing at stand-up, I've update the spec with the proposed options. I'm suggesting we use a basic sort initially, then add an enumerated sort as a follow-on PR.

Reviewer 2: The spec does not seem to reflect your comment. It seems to just have verbose language trying to justify why any other sort option is bad instead of choosing basic sort today and sorting via dom ordering of mapping elements as future work.

Reviewer 2: This issue is this commit wasn't showing up for me: https://github.com/ni/nimble/pull/1220/commits/7e27c21438950947e7deaf9fc2682fb8d3079b80
Maybe was accidentally not pushed or I needed to refresh or something.
But that commit and the follow-on of https://github.com/ni/nimble/pull/1220/commits/643cd8f5cc4a87fe01f5d91e495958c94e937e98 together look great.

> For icons, if multiple values map the to same icon, it is possible that sorting will result in the instances of a certain icon not being all together in one span of rows. The user may think they should be all together, but this is a corner case that we can't/shouldn't do anything about.

Reviewer 2: ```suggestion
For icons, if multiple values map the to same icon, it is possible that sorting will result in the instances of a certain icon not being all together in one span of rows. Users will be expected to provide visually distinct icons for each mapping as the column will not enforce or validate distinct icons for each mapping.
```

### Angular integration

Angular directives will be created for the column components and the mapping components. No component has form association, so a `ControlValueAccessor` will not be created.

### Blazor integration

Blazor wrappers will be created for the components. Columns will be generic in the type of the key, and will cascade that type parameter to contained mapping elements (see [`CascadingTypeParameter`](https://learn.microsoft.com/en-us/aspnet/core/blazor/components/generic-type-support?view=aspnetcore-7.0#cascaded-generic-type-support)):

```HTML
<NimbleTableColumnMapping TKey=int Field="NumberData">
    <NimbleMappingText Key="1" Label="foo"></NimbleMappingText>
    <NimbleMappingText Key="2" Label="bar"></NimbleMappingText>
</NimbleTableColumnMapping>
```

### Visual Appearance

> The cell view (and group header view) will be responsible for styling the templates returned by the mappings. This will include alignment and spacing (`--ni-nimble-small-padding`). Unfortunately, because the mappings cannot provide CSS to go with the templates they return, some implementation knowledge will leak from the mappings to the cell/group header views. For example, we must set `flex-shrink: 0` on all elements so that icons do not get smaller in a group header when the adjacent label takes up all the space.

Reviewer 2: I don't think I understand the concerns being expressed here but it also feels like implementation details that could be discussed in review. If you feel like it should have additional discussion to capture in the HLD to avoid PR churn we can discuss it.

Author: I confused Molly with this same concern. Maybe we don't need to discuss this implementation detail until the actual implementation PR.

Reviewer: Can we remove the last part of this paragraph then? I think the first part makes sense, but if we all agree that the leaking of implementation details is an implementation detail, then I'd vote for taking it out of this document.

Author: Sure. Removed.

# Mapping Table Column

## Overview

The `nimble-table-column-mapping` is a component that supports rendering specific number, boolean, or string values as mapped, predefined, html fragments (with content restrictions).

### Background

[Icon column type issue](https://github.com/ni/nimble/issues/1013)

[Boolean column type issue](https://github.com/ni/nimble/issues/1103)

### Features

-   Supported input:
    -   string
    -   number
    -   boolean
-   Supported output:
    -   Text
    -   One or more Nimble icons (with colors via icon severity property)
    -   Nimble spinner
    -   One span of text preceded and/or followed by icons/spinner
    -   (empty)

### Non-goals

-   Non-Nimble icon support
-   Arbitrary icon colors
-   Hyperlink icons
-   Non-icon, non-spinner Nimble components
-   Arbitrary html

---

## Design

> In addition to the `nimble-table-column-mapping` element, there are `nimble-mapping-<input type>` elements to define mappings between values and html fragments. Below is an example of how these elements would be used within a `nimble-table`:

Reviewer 2: It's not clear to me the benefit of having multiple mapping elements based on input type.

A column can't mix types, it's either number|null, string|null, boolean|null, etc. So it should be defined per column, not per mapping entry.

So why define the elements with the flexibility that they can mix multiple representations within a column? Just makes for extra validation work.

Could `nimble-table-column-mapping` have a type property and then have a single mapping element of `key: type boolean | string | number`?

Author: Taking suggested approach

</nimble-table>
```

Note that the `key` attribute values are always given as strings. In the case of boolean or number mappings, this value is parsed to the appropriate type, so that it can properly be compared to the values from the row records.

If a template includes text, that text should be within a `span` element, otherwise, when the text is too long for the cell, it it will not be ellipsized and there will be no tooltip to show the full value.

```TS
import {
    attr,
    booleanConverter,
    nullableNumberConverter
} from '@microsoft/fast-element';

export class TableColumnNumberMapping extends TableColumnMapping {
    @attr({ converter: nullableNumberConverter })
    public override key: number | null = null;
}

export class TableColumnBooleanMapping extends TableColumnMapping {
    @attr({ converter: booleanConverter })
    public override key: boolean | null = null;
}

```

When none of the given mappings match the record value for a cell, that cell will be empty. Alternatively, we could provide support for a default mapping (i.e. `nimble-mapping-default`) that would be rendered when no other values matched. However this doesn't seem worth the effort/complexity.

> **Alternatives:**

Reviewer 2: Can you discuss the alternative of not slotting elements but instead just exposing configuration via attributes?

In concert with the above proposal where the type of the element is not part of the name, we can then make these represent the visual:

```
nimble-mapping-text key="blah" text="blah"
nimble-mapping-icon key="blah" icon="nimble-icon-blah" title="blah"
nimble-mapping-text-icon key="blah" icon="nimble-icon-blah" title="blah" text="blah"
nimble-mapping-spinner key="blah"
```

I'd like to see a discussion of the pros and cons. I think in this case a big con is giving too much flexibility in the API via the template elements. It's very verbose for the user and I think we would want a lot of validation to be strict about only providing a single icon or a single span or a single version of both in a specific order. I think any flexibility there is a pain.

Reviewer 3: +1

Author: After our group discussion of this, I've updated the spec to use this non-template-based approach.

### Anatomy

#### `nimble-table-column-mapping`

```HTML
<template slot="${x => x.columnInternals.uniqueId}">
    <slot
        ${slotted('mappings')}
        name="mapping">
    </slot>
    <span class="header-content">
        <slot></slot>
    </span>
</template>
```

Cell view:

```HTML
<div
    ${ref('wrapper')}
    class="wrapper"
    @mouseover="${x => {
        x.isValidContentAndHasOverflow = !!x.wrapper.innerText && x.wrapper.offsetWidth < x.wrapper.scrollWidth;
    }}"
    @mouseout="${x => {
        x.isValidContentAndHasOverflow = false;
    }}"
    title=${x => (x.isValidContentAndHasOverflow ? x.wrapper.innerText : null)}
>
    ${repeat(x => (x.column as TableColumnMapping).mappings,
        html<TableColumnMapping, TableColumnMappingCellView>`
            ${when((x, c) => x.key === (c.parent as TableColumnMappingCellView).cellRecord.value,
                html<TableColumnMapping>`
                    ${repeat(x => x.mappedTemplate,
                        html<HTMLTemplateElement>`${x => new ViewTemplate(x.innerHTML, [])}`
                    )}
                `
            )}
        `
    )}
</div>
```

> Header view:

Author: Fixed

> Note the following requires that `TableColumnMappingHeaderView` has a reference to the column with which it is associated. This is needed to enumerate the column's mapping elements.

Reviewer 3: ```suggestion
Note the following requires that `TableColumnMappingGroupHeaderView` has a reference to the column with which it is associated. This is needed to enumerate the column's mapping elements.
```

Author: Fixed

                    )}
                `
            )}
        `
    )}
</div>
```

For each mapping, if the key matches the value for this row, get the mapping's contained `template` and render its `innerHTML`.

#### `nimble-mapping-string/number/boolean`

```HTML
<template slot="mapping">
    <slot
        ${slotted({ property: 'mappedTemplate', filter: x => x instanceof HTMLTemplateElement })}
    >
    </slot>
</template>
```

### Grouping

Will support grouping by the record value. The grouping header will display the rendered html mapped to that value. If multiple different values map to the same rendered html (e.g. icon), there may be separate groups that visually look like they should be combined into one group. We could instead render the record value into the header, but that value is not otherwise exposed to the user. It may not even have much meaning (e.g. some numeric code). I think it is more reasonable to display the rendered html.

### Sorting

Sorting has the same issue as grouping, where visually identical cells could end up not sorted contiguously. Again, I believe this is reasonable.

### Sizing

> We will support fixed or fractional widths for this column type. We will introduce a new mixin for fixed-width support that exposes a `pixel-width` property. If this is set, the column will have a fixed width, otherwise it defaults to fractional width. There is an undocumented minimum column width, so a client may set `pixel-width` to 1 to end up with that minimum width.

Reviewer 3: I'm not sure I understand the intent here. When you say "we will support fixed _or_ fractional widths", are you suggesting that the user be able to configure the fractional width setting (I would expect this)? If so, then there should not be an "undocumented" minimum column width, that comes with the `FractionalWidthColumn` mixin. Additionally, this would suggest that the API (described above) should include `fractional-width` and `minimum-pixel-width`.

If the intent is to _not_ allow fractional width adjustment, then I think we should have it behave in such a way that it _never_ behaves in such a way. This would require us to provide a default for `pixel-width` that is non-null/undefined, and also handle the case in the mixin where a user sets it to null/undefined and have it reset to the provided default (since a pixel-width column must have a non-null/undefined value in order to not be sized fractionally).

Author: Updated the spec to describe new sizing plan.

### Angular integration

Angular directives will be created for the column component and the mapping components. The component will not have form association, so a `ControlValueAccessor` will not be created.

### Blazor integration

> A Blazor wrapper will be created for the component. _Initial attempts at prototyping resulted in exceptions being thrown and no icons being rendered. This is an open issue._

Reviewer 3: For me this is a stronger indication that we need to question the API design being proposed. It's possible that the use of the `template` element is untenable in Blazor (for whatever reason), and if so, I don't think we could move forward with that as our design.

Reviewer 2: +1  the template stuff is a new pattern not vetted yet, investigating sooner could prevent a lot of rework

Author: Figued out my issue. It seems Blazor elements aren't supported inside `template` elements. The result is that the `innerHTML`/`content` of the `HTMLTemplateElement` is empty. FAST chokes on the empty `ViewTemplate` I was creating, and that was the exception I was getting. Is it reasonable to direct Blazor clients to provide their mapped content using the Nimble elements rather than their Blazor wrappers?

Reviewer 3: I think this is a reasonable suggestion. Particularly since we aren't expecting clients to bind aspects of the elements within the `template` to something on the parent component. They lose type safety for the attributes they wish to set on the components, but this may just be the cost of doing business in Blazor. Not sure if we can file an issue to ASP.NET Core asking for this to be fixed (this seems pretty corner case).

Author: Moot now with switch away from using `template` elements.

```HTML
    <nimble-mapping-text key="NOT_STARTED" label="Not Started"></nimble-mapping-text>
    <nimble-mapping-text key="RUNNING" label="Running"></nimble-mapping-text>
    <nimble-mapping-text key="DONE" label="Done"></nimble-mapping-text>
```

We would need new support for sorting this way. We could define a new sorting option, "enumerated sort", where the column provides an ordered list of values, and the table sorts the column based on that given order. To compare the relative order of two values, we at have to search the list, making the sort operation a bit more expensive, but still probably reasonable except in the case of enums with many values.

> Our options are to use a basic sort (for all three value types), use an enumerated sort (again, for all three value types), or to start with a basic sort and implement the enumerated sort in a follow-on PR. Settling for a basic sort is reasonable and the easiest solution, but it would not give as nice an experience as an enumerated sort. The enumerated sort would expand the scope of this feature and could easily be implemented as a standalone change. For those reasons, I propose we use a basic sort for the initial submission of the column type, but we have a follow-on PR to add the enumerated sort and adopt that for the mapping/icon column.

Reviewer: This is a bit nit-picky, but I don't think the terminology of saying something would be done in a "follow-on PR" is quite right. Since this is an HLD, its goal isn't to capture how many PRs will be involved in completing this work. I think the more accurate thing would be to say something like "this could be done as follow-on work that is out of scope for the initial implementation of these components".

Author: I've changed the wording a bit so I don't reference "PRs."

> For icons, if multiple values map the to same icon, it is possible that sorting will result in the instances of a certain icon not being all together in one span of rows. Users will be expected to provide visually distinct icons for each mapping as the column will not enforce or validate distinct icons for each mapping.

Reviewer 3: ```suggestion
For icons, if multiple values map to the same icon, it is possible that sorting will result in the instances of a certain icon not being all together in one span of rows. Users will be expected to provide visually distinct icons for each mapping as the column will not enforce or validate distinct icons for each mapping.
```

### Globalization

All text will be provided by the client and is expected to be localized.

### Security

N/A

### Performance

N/A

### Dependencies

None

### Test Plan

-   Unit tests will be written verifying the usual component expectations, plus:
    -   renders mapping matching the cell value (string, number, and boolean)
    -   nothing rendered when value matches no mappings
    -   validation error when non-unique mapping keys exist
    -   validation error when multiple mappings marked as default
    -   validation error when mapping key is null and not marked default
    -   validation error when invalid icon name given
    -   validation error when icon column has a `nimble-mapping-text` element
    -   grouping header for icon column includes label
-   Verify manually that the column content appears in the accessibility tree and can be read by a screen reader.
-   Verify manually that several mapping columns with thousands of elements scrolls performantly.
-   Visual Chromatic tests will be created

### Tooling

N/A

### Documentation

Documented in Storybook

---

> ## Open Issues

Reviewer: I thought there were still some open questions on the sizing modes we thought the icon-only column needed. Is that not the case? Are we set on having it only support fixed pixel widths?

Author: As far as I know (and the spec says), the icon column will be fixed width and the mapping column will use either fractional width or fixed width. No one else has commented about it, so I take that to mean we're good with that. Do you think we need to do something different?

        <nimble-mapping-icon key="error" icon="nimble-icon-xmark" severity="error" label="Errored"></nimble-mapping-icon>
        <nimble-mapping-icon key="pass" icon="nimble-icon-check" severity="success" label="Passed"></nimble-mapping-icon>
        <nimble-mapping-spinner key="running" label="Running"></nimble-mapping-spinner>
    </nimble-table-column-icon>
    <nimble-table-column-mapping field-name="errorCode" data-type="number">
        Error Code
        <nimble-mapping-text key="1" label="A bad thing happened"></nimble-mapping-text>
        <nimble-mapping-text key="2" label="A worse thing happened"></nimble-mapping-text>
        <nimble-mapping-text key="3" label="A terrible thing happened"></nimble-mapping-text>
    </nimble-table-column-mapping>
    <nimble-table-column-icon field-name="archived" data-type="boolean">
        Archived
        <nimble-mapping-icon key="true" icon="nimble-icon-database" label="Archived"></nimble-mapping-icon>
    </nimble-table-column-icon>
</nimble-table>
```

Each column contains mapping elements that define what to render when the cell's value matches the given `key` value.

> When none of the given mappings match the record value for a cell, that cell will be empty. Alternatively, if one of the mappings has the `default` attribute, it will match when no other mappings have. This is equivalent to the `placeholder` configuration we provide on `nimble-table-column-text` and `nimble-table-column-anchor`.

Reviewer 2: Can we use a name more specific than default? Like if someone asked "It's the default what?", I think it would be something like:

- default-mapping
- fallback-mapping
- default-key
- fallback-key

Also I guess key being allowed to be null is so you can specify a default-mapping without a key. So maybe in the validation section we need a case for the key is null but not marked as the default-mapping (i.e. a null key must be marked as default-mapping).

Author: >  if someone asked "It's the default what?"

I would also reply, "the default mapping". Since the **element is the mapping**, it seems like making the attribute name `default-mapping` is redundant. I like having it just be `default`, just like the `default` case in a switch statement.

I've added a test case for the one you pointed out.

Author: Ok, can't argue with that. I'll go with `default-mapping`.

> Validation will be performed to ensure each mapping's key value can be converted to the `data-type` of the column. If not, an error flag will be set on the column's validation object. Note that whenever an error flag is set on the column's validation object, a generic `invalidColumnConfiguration` flag is also set on the table, putting it in an invalid state as well.

Reviewer 2: I think we should go with the name `key-type`, better associates for users that it maps to the key attribute below

Author: Ok

**Alternatives:**

An earlier version of this spec proposed mapping elements with `template` elements as content instead of relying on attribute configuration. The template element would define the mapped html to render. We would impose restrictions on the types of supported elements that could be provided in the template.

Pros:

-   It would not require updates to the API if we needed to support new types of mapped content (e.g. icon with text), or if the mapped content itself got new configuration options (e.g. a scaling factor for icons).
-   Undefined element types caught at compile time.

Cons:

-   Verbose. Requires user to create `template` element and wrap text in `span`s for styling purposes.
-   Requires difficult validation to ensure only supported elements are present in the `template`.
-   Could allow users to provide inline styling.
-   Blazor: cannot put Blazor components inside `template`--must use raw Nimble elements without type safety

### API

#### Icon column element:

_Component Name_

-   `nimble-table-column-icon`

_Props/Attrs_

-   `field-name`: string

> -   `data-type`: 'string' | 'number' | 'boolean' | null (defaults to 'string')

Reviewer 4: What would a `null` value for `data-type` mean? If it always results in an error, is it possible to disallow that value?

Reviewer 2: In the current spec it shows a null type referring to a type of string, but also that "string" is a possible value.
I think we should only have one representation of each format, either:
 - undefined/null (default: string) / "number" / "boolean" (2 keys and a default)
 - "string" / "number" / "boolean" and no default value so error if null /undefined (3 keys and no default)

I don't feel strongly about which of those options we choose. In general I prefer good defaults (the 2 keys option) but in this case maybe choosing a type by default could lead to subtle bugs. For example I forget to specify the type and pass numbers as strings which are then treated as strings in comparisons of "1" and "1.0" and "+1.0" and "1e0", etc.

So in this case I think I prefer 3 keys no default, i.e. where Jesse is going.

Reviewer 4: Good point, it's a tradeoff between:
- 2 keys and a default being easier to configure (many cases won't need to specify `data-type` at all)
- 3 keys and no default being less likely to introduce subtle bugs because clients must always specify `data-type` so presumably they'd also think about its value

I'm good with 3 keys no default.

Author: Updated spec

-   `field-name`: string
-   `key-type`: 'string' | 'number' | 'boolean'
-   `pixel-width`: number (set to the desired fixed column width, else will use a default fixed width)

_Content_

-   column title (text or icon)
-   1 or more `nimble-mapping-icon` or `nimble-mapping-spinner` elements

#### General mapping column element:

_Component Name_

-   `nimble-table-column-mapping`

_Props/Attrs_

-   `field-name`: string
-   `key-type`: 'string' | 'number' | 'boolean'
-   `fractional-width`: number (defaults to 1)
-   `min-pixel-width`: number (defaults to minimum supported by table)

_Content_

-   column title (text or icon)
-   1 or more `nimble-mapping-text` elements

#### Mapping element (icon):

_Component Name_

-   `nimble-mapping-icon`

_Props/Attrs_

> -   `key`: string | number | boolean | null

Reviewer: I might be missing something here, but what is the purpose of `null` here? Just that it's valid if the client doesn't specify a `key` on their mapping? Should this be `undefined` instead? That's typically the value we use to indicate that a property doesn't have a value.

Author: Yes, that was the intent. I'll make it undefined instead.

-   `label`: string - localized value used as the accessible name and `title` of the icon. Will also be displayed in the group header.
-   `default`: boolean - presence causes this mapping to be used when no others match the value

#### Mapping element (spinner):

_Component Name_

-   `nimble-mapping-spinner`

_Props/Attrs_

-   `key`: string | number | boolean | null
-   `label`: string - localized value used as the accessible name and `title` of the spinner. Will also be displayed in the group header.
-   `default`: boolean - presence causes this mapping to be used when no others match the value

#### Mapping element (text):

_Component Name_

-   `nimble-mapping-text`

_Props/Attrs_

-   `key`: string | number | boolean | null
-   `label`: string - display text
-   `default`: boolean - presence causes this mapping to be used when no others match the value

### Anatomy

> #### `nimble-table-column-mapping`

Reviewer 4: A couple thoughts:
1. this will have the logic to display a tooltip when text overflows, right? It's not necessary to copy it again in the spec, but would be good to call it out
2. hopefully this can reuse some of the components described in #1224. A goal of that HLD is that anything that renders as text (text columns, numeric columns, date time columns, etc) will render identically by sharing base classes / cell elements / header elements. I think we'd want that to apply to mapping columns that display text too.

Author: 1. I didn't put the templates for the mapping text element in this section, but I did call this out in the design section:
> Text in a grouping header or in the cell will be ellipsized and gain a tooltip when the full text is too long to display.

2. I'll try to share whatever I can.

#### `nimble-table-column-mapping`

```HTML
<template slot="${x => x.columnInternals.uniqueId}">
    <slot
        ${slotted('mappings')}
        name="mapping">
    </slot>
    <span class="header-content">
        <slot></slot>
    </span>
</template>
```

Cell view:

The cell view relies on the matched mapping to provide a template to render.

```HTML

> <div ${ref('wrapper') class="wrapper"}>

Reviewer: This might be more of an implementation-PR discussion, but I'm curious why all of this logic is being written in the template. I think it might be more straightforward for the cell view to have a getter that returns the matching mapping that should get rendered in the template. I could be wrong, but that seems like it would be more efficient that having the template have a bunch of conditional logic that's dependent on the state of other parts of the template.

Author: That's a fair point. Somehow, I have a bias for doing things in the template, but I don't think I have a justification.

    ${repeat(x => x.columnConfig.mappings,
        html<Mapping, TableColumnMappingCellView>`
            ${when((x, c) => x.key === (c.parent as TableColumnMappingCellView).cellRecord.value,
                html<Mapping>`${x => x.cellViewTemplate}`
            )}
        `
    )}
</div>
${when(x => x.wrapper.childElementCount === 0, html<TableColumnMappingCellView>`
    ${repeat(x => x.columnConfig.mappings,
        html<Mapping, TableColumnMappingCellView>`
            ${when((x, c) => x.defaultMapping,
                html<Mapping>`${x => x.cellViewTemplate}`
            )}
        `
    )}
`)}
```

Group header view:

Similarly, the group header view relies on the matched mapping to provide a template to render.

> ```HTML

Reviewer: Same here. It seems like part of this conditional logic should really belongs in the group header view class, not the template.

> Note the following requires that `TableColumnMappingGroupHeaderView` has a reference to the column with which it is associated. This is needed to enumerate the column's mapping elements.

Reviewer 4: Without looking at the code I'm not sure if this is a good suggestion or not, but did you consider inverting this dependency? Instead of having the header view query the column to find out which mapping to display, make it the column's responsibility to set attributes on each header view so that it has the information it needs to render.

I can think of a couple ways to achieve this:
1. put that information in column internals `columnConfig`, which is generic and is already provided to the header view
2. change `createGroupHeaderViewTemplate()` to be a virtual method on a column instead of a static helper. Then the mapping columns could override it to add additional configuration.

Just some ideas, it might require prototyping to decide if they're feasible and whether we like them better. It just feels wrong for an element to be looking up information in other elements in order to decide what to render.

I would be ok with deferring this discussion until implementation if it's easier to have when referring to code.

Author: I've updated the spec accordingly.

    ${x => x.cellRecord.value)}}
`)}
```

#### `nimble-mapping-*`

```HTML
<template slot="mapping"></template>
```

#### `nimble-mapping-icon`

`mapping.cellViewTemplate`:

```HTML
<${this.icon}
    title="${x => x.label}"
    aria-label="${x => x.label}"
    severity="${x => x.severity}">
</${this.icon}>
```

`mapping.groupHeaderViewTemplate`:

```HTML
<${this.icon}
    title="${x => x.label}"
    aria-label="${x => x.label}"
    severity="${x => x.severity}">
</${this.icon}>
<span
    ${ref('span')}
    @mouseover="${x => {
        x.isValidContentAndHasOverflow = !!x.label && x.span!.offsetWidth < x.span!.scrollWidth;
    }}"
    @mouseout="${x => {
        x.isValidContentAndHasOverflow = false;
    }}"
    title=${x => (x.isValidContentAndHasOverflow ? x.label : null)}>
    ${x => x.label}
</span>`;
```

### Grouping

Grouping will be based on the record value. The grouping header will display the rendered icon/spinner/text. In the case of an icon/spinner, it will also be followed by the `label` text. This will disambiguate cases where multiple record values map to the same icon (assuming the labels are different).

> For values that do not match any mapping, we will display the raw data value. While this introduces inconsistency, it seems preferable to the alternative, which is having multiple, separate groupings with a blank header (well, with just the item count in parens). Even in the case where there is a default mapping, we would still end up with separate groups with the identical default mapped icon and/or text, which is just a bad.

Reviewer: ```suggestion
For values that do not match any mapping, we will display the raw data value. While this introduces inconsistency, it seems preferable to the alternative, which is having multiple, separate groupings with a blank header (well, with just the item count in parens). Even in the case where there is a default mapping, we would still end up with separate groups with the identical default mapped icon and/or text, which is just as bad.
```

Author: fixed

### Sizing

`nimble-table-column-icon` will support only a fixed width. We will introduce a new mixin for fixed-width support that exposes a `pixel-width` property. The default value will be the minimum supported by the table, which is still significantly larger than the width of an icon.

`nimble-table-column-mapping` will support fixed or fractional widths. If `pixel-width` is set, the column will have a fixed width, otherwise it defaults to a fractional width of 1. The client may configure `fractional-width` and/or `min-pixel-width`.

### Angular integration

Angular directives will be created for the column components and the mapping components. No component has form association, so a `ControlValueAccessor` will not be created.

### Blazor integration

Blazor wrappers will be created for the components. Mapping components will be generic in the type of the key:

```HTML
<NimbleTableColumnMapping Field="NumberData" DataType="number">
    <NimbleMappingText TKey=int Key="1" Label="foo"></NimbleMappingText>

>     <NimbleMappingText TKey=int Key="2" Label="bar"></NimbleMappingText>

Reviewer 3: Is this generic API useful now? If it were always a "string" wouldn't the underlying `nimble-table-column-mapping` still do the right thing because its `datatype` is set to `number`? As is, it seems like we've opened the door for a mismatch of types between the `NimbleTableColumnMapping` and its contained mapping elements.

Author: I've updated the spec to indicate we'll be using `CascadingTypeParameter`.
